python/dgl/backend/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,356 @@
+import abc
+from abc import abstractmethod
+import torch as th
+
+from .util import get_shared_mem_array, create_shared_mem_array
+from .sparse_emb import NodeEmbedding
+
+class SparseGradOptimizer(abc.ABC):
+    r''' The abstract sparse optimizer.
+
+    Note: dgl sparse optimizer only work with dgl.NodeEmbedding
+
+    Parameters
+    ----------
+    params : list of NodeEmbedding
+        The list of NodeEmbeddings.
+    lr : float
+        The learning rate.
+    '''
+    def __init__(self, params, lr):
+        self._params = params
+        self._lr = lr
+        self._rank = None
+        self._world_size = None
+        self._shared_cache = {}
+        self._clean_grad = False
+
+    def step(self):
+        ''' The step function.
+
+        The step function is invoked at the end of every batch to update embeddings
+        '''
+        with th.no_grad():
+            # Frequently alloc and free shared memory to hold intermediate tensor is expensive
+            # We cache shared memory buffers in shared_emb.
+            shared_emb = {emb.name: ([],[]) for emb in self._params}
+
+            # Go through all sparse embeddings
+            for emb in self._params:
+                num_embeddings = emb.num_embeddings
+                emb_name = emb.name
+
+                # Each gpu process takes the resposibility of update a range of sparse embedding,
+                # thus we can parallel the gradient update.
+                range_size = (num_embeddings + self._world_size - 1) // self._world_size \
+                    if self._world_size > 0 else 0
+                for idx, data in emb._trace:
+                    grad = data.grad.data
+                    device = grad.device
+                    idx_dtype = idx.dtype
+                    grad_dtype = grad.dtype
+                    grad_dim = grad.shape[1]
+
+                    if self._world_size > 0:
+                        if emb_name not in self._shared_cache:
```

Comment:
what is `_shared_cache` used for?

Reply:
This is a optimization. It will cache the shared-memory used by sending gradients and reuse it next time. So there will not be lots of memory alloc and free.